ui2/src/context/ServicesContext.tsx

```diff
@@ -0,0 +1,130 @@
+type Service = {
+  contract: ServiceContract;
+  service: ServiceKind;
+}
+
+type ServicesState = {
+  services: Service[];
+  loading: boolean;
+}
+
+const ServicesStateContext = React.createContext<ServicesState>({ services: [], loading: false });
+
+const ServicesProvider : React.FC = ({ children }) => {
+  const [ services, setServices ] = useState<Service[]>([]);
+  const [ loading, setLoading ] = useState<boolean>(false);
+
+  const { contracts: custodyService, loading: custodyLoading } = useStreamQueries(CustodyService);
+  const { contracts: auctionService, loading: auctionLoading } = useStreamQueries(AuctionService);
+  const { contracts: biddingService, loading: biddingLoading } = useStreamQueries(BiddingService);
+  const { contracts: issuanceService, loading: issuanceLoading } = useStreamQueries(IssuanceService);
+  const { contracts: listingService, loading: listingLoading } = useStreamQueries(ListingService);
+  const { contracts: tradingService, loading: tradingLoading } = useStreamQueries(TradingService);
+
+  useEffect(() =>
+    setLoading(
+      custodyLoading  ||
+      auctionLoading  ||
+      biddingLoading  ||
+      issuanceLoading ||
+      listingLoading  ||
+      tradingLoading
+    ), [custodyLoading, auctionLoading, biddingLoading, issuanceLoading, listingLoading, tradingLoading]
+  );
+
+  useEffect(() =>
+    setServices([
+      ...custodyService.map(c => ({ contract: c, service: ServiceKind.CUSTODY })),
+      ...auctionService.map(c => ({ contract: c, service: ServiceKind.AUCTION })),
+      ...biddingService.map(c => ({ contract: c, service: ServiceKind.BIDDING })),
+      ...issuanceService.map(c => ({ contract: c, service: ServiceKind.ISSUANCE })),
+      ...listingService.map(c => ({ contract: c, service: ServiceKind.LISTING })),
+      ...tradingService.map(c => ({ contract: c, service: ServiceKind.TRADING }))
+    ]), [custodyService, auctionService, biddingService, issuanceService, listingService, tradingService]
+  );
+
+  return (
+    <ServicesStateContext.Provider value={{services, loading}}>
+      {children}
+    </ServicesStateContext.Provider>
+  );
+}
+
+type GroupedCustomerServices = {
+  provider: string;
+  services: ServiceKind[];
+  contracts: ServiceContract[]
+}[];
+
+/* Retrieve all providers who are providing you, the customer, a service */
+function useProviderServices (party: string): GroupedCustomerServices {
+  const context = React.useContext<ServicesState>(ServicesStateContext);
+  if (context === undefined) {
+    throw new Error("useProviderServices  must be used within a ServicesProvider");
+  }
+
+  return context.services
+    .filter(s => s.contract.payload.customer === party)
+    .reduce((acc, service) => {
+      const providerDetails = acc.find(i => i.provider === service.contract.payload.provider);
+
+      const provider = providerDetails?.provider || service.contract.payload.provider;
+      const services = [ ...(providerDetails?.services || []), service.service];
```

Comment:
You could use `_.compact(providerDetails.services)` from lodash which matches the pattern above where you return an empty list if an object is undefined, null, etc.

See https://lodash.com/docs/4.17.15#compact.

Reply:
This is awesome, thanks!